<a href="https://colab.research.google.com/github/ShubhamCJagtap/Capstone_Project_1_EDA/blob/master/Capstone_project_1_Mobile_Events_EDA_colab_notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

 ## EDA steps to follow
 
**1.** [**Introduction**](#Section1)<br>
**2.** [**Problem Statement**](#Section2)<br>
**3.** [**Installing & Importing Libraries**](#Section3)<br>
  - **3.1** [**Installing Libraries**](#Section31)
  - **3.2** [**Upgrading Libraries**](#Section32)
  - **3.3** [**Importing Libraries**](#Section33)

**4.** [**Data Acquisition & Description**](#Section4)<br>
  - **4.1** [**Data Description**](#Section41)
  - **4.2** [**Data Information**](#Section42)

**5.** [**Data Pre-profiling**](#Section5)<br>
**6.** [**Data Cleaning**](#Section6)<br>
**7.** [**Data Post-profiling**](#Section7)<br>
**8.** [**Exploratory Data Analysis**](#Section8)<br>
  - **8.1** [**Outlet Metrics-based Analysis**](#Section81)
  - **8.2** [**Nutritional Value-based Analysis**](#Section82)
  - **8.3** [**Geographical Information-based Analysis**](#Section83)
  - **8.4** [**Outlet Metrics-based & Nutritional Value-based Analysis**](#Section84)
  - **8.5** [**Outlet Metrics-based & Geographical Information-based Analysis**](#Section85)
  - **8.6** [**Nutritional Value-based & Geographical Information-based Analysis**](#Section86)
  - **8.7** [**Outlet Metric-based, Menu Items & Geographical Information-based Analysis**](#Section87)

**9.** [**Summarization**](#Section9)<br>

# Introduction

### Company Introduction
Your company for this project is InsaidTelecom.<br>

InsaidTelecom, one of the leading telecom players, understands that customizing offerings is very important for its business to stay competitive.<br>
Currently, InsaidTelecom is seeking to leverage behavioural data from more than 60% of the 50 million mobile devices active daily in India<br>
They are doing this to help their clients better understand and interact with their audiences.<br>


# problem Statement

## Current Scenario
In this consulting assignment, Insaidians are expected to build a dashboard.<br>
This dashboard will help us to understand a user's demographic characteristics based on their mobile usage, geolocation, and mobile device properties<br>
Doing so will help millions of developers and brand advertisers around the world pursue data-driven marketing efforts.
These efforts are relevant to their users and cater to their preferences.<br>

### Expected output

- EDA observations
- Dashboard

# Role

## Your Role
- To help the customer the consultants are expected to have depth of clarity in the underlying data.
- How much effort has been put into cleansing and purifying the data will decide how closely have you looked at the data..
- How detailed is the observation stated in the submission report and finally, how well a group presents their consulting journey.
- Please remember that this is an analytics consulting firm. Hence, your efforts in terms of finding user behaviour are going to directly impact the company's offerings.
- Do help the company understand what is the right way forward and suggest actionable insights from marketing and product terms.


### State to focus
- Delhi

### Provided Instructions

#### Dataset Description
- In this assignment, you are going to study the demographics of a user (gender and age) based on their app download and usage behaviours.
- The Data is collected from mobile apps that use Inside Telecom services. Full recognition and consent from an individual user of those apps have been obtained,and appropriate anonymization has been performed to protect privacy. Due to confidentiality, we won't provide details on how the gender and age data were obtained.
- Please treat them as accurate ground truth for prediction.


## Week 1 Task

- Fetch the data on to the notebook and ponder upon the challenges present in the data
- All observations supposed to be listed in observation sheet

# Import Libraries

In [8]:
!pip install mysql.connector

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [9]:
import pandas as pd
import numpy as np
import mysql.connector as connection # install !pip install mysql.connector
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

pd.set_option('display.max_columns',None)
pd.set_option('display.max_rows',None)
pd.set_option('display.max_colwidth',None)
pd.set_option('mode.chained_assignment',None)

In [10]:
Google_drive_link = 'https://drive.google.com/file/d/1Ir3rW0YTKmk7MSjVjCU_UGMQevhe1v9W/view'

# Collect Events Data

In [11]:
df = pd.read_csv('/content/drive/MyDrive/Capstone 1 Insaid/Database/events_data.csv')   #read csv file

In [12]:
df.head(10)

event_id     device_id            timestamp  longitude   latitude  \
0   2765368  2.973348e+18  2016-05-07 22:52:05  77.225676  28.730140   
1   2955066  4.734221e+18  2016-05-01 20:44:16  88.388361  22.660325   
2    605968 -3.264500e+18  2016-05-02 14:23:04  77.256809  28.757906   
3    448114  5.731369e+18  2016-05-03 13:21:16  80.343613  13.153332   
4    665740  3.388880e+17  2016-05-06 03:51:05  85.997745  23.842609   
5   1078723 -5.124242e+17  2016-05-02 02:21:20  83.398244  17.768149   
6   2948755  5.536513e+18  2016-05-03 16:20:14  80.359796  13.128174   
7    280014 -8.879644e+18  2016-05-05 13:06:01  78.155397  16.390327   
8   2309828  3.020769e+18  2016-05-04 09:24:50  77.263922  28.757914   
9    720049  4.928556e+18  2016-05-06 19:00:52  72.844795  18.974929   

            city          state  
0          Delhi          Delhi  
1       Calcutta     WestBengal  
2          Delhi          Delhi  
3        Chennai      TamilNadu  
4         Bokaro      Jharkhand  
5  Visakhapatnam  AndhraPradesh  
6        Chennai      TamilNadu  
7       Wanparti  AndhraPradesh  
8          Delhi          Delhi  
9         Mumbai    Maharashtra

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3252950 entries, 0 to 3252949
Data columns (total 7 columns):
 #   Column     Dtype  
---  ------     -----  
 0   event_id   int64  
 1   device_id  float64
 2   timestamp  object 
 3   longitude  float64
 4   latitude   float64
 5   city       object 
 6   state      object 
dtypes: float64(3), int64(1), object(3)
memory usage: 173.7+ MB


## Observations (data.info())

- Total Entries  3252950
- timestamp type object ## need to change in datetime dtype

In [14]:
df.describe()

event_id     device_id     longitude      latitude
count  3.252950e+06  3.252497e+06  3.252527e+06  3.252527e+06
mean   1.626476e+06  1.012200e+17  7.815868e+01  2.168851e+01
std    9.390459e+05  5.316758e+18  4.235570e+00  5.789111e+00
min    1.000000e+00 -9.222957e+18  1.256700e+01  8.190110e+00
25%    8.132382e+05 -4.540611e+18  7.583543e+01  1.780171e+01
50%    1.626476e+06  1.726820e+17  7.726814e+01  2.216454e+01
75%    2.439713e+06  4.861813e+18  8.031916e+01  2.868278e+01
max    3.252950e+06  9.222849e+18  9.545946e+01  4.187190e+01

In [15]:
describe_df = df.describe()

In [16]:
describe_df[describe_df.index=='count']

event_id  device_id  longitude   latitude
count  3252950.0  3252497.0  3252527.0  3252527.0

In [17]:
describe_df

event_id     device_id     longitude      latitude
count  3.252950e+06  3.252497e+06  3.252527e+06  3.252527e+06
mean   1.626476e+06  1.012200e+17  7.815868e+01  2.168851e+01
std    9.390459e+05  5.316758e+18  4.235570e+00  5.789111e+00
min    1.000000e+00 -9.222957e+18  1.256700e+01  8.190110e+00
25%    8.132382e+05 -4.540611e+18  7.583543e+01  1.780171e+01
50%    1.626476e+06  1.726820e+17  7.726814e+01  2.216454e+01
75%    2.439713e+06  4.861813e+18  8.031916e+01  2.868278e+01
max    3.252950e+06  9.222849e+18  9.545946e+01  4.187190e+01

In [18]:
a = describe_df[describe_df.index=='count']['event_id']
for i in describe_df[describe_df.index=='count'].columns:
    print("Missing Values in {} are {}".format(i,a - describe_df[describe_df.index=='count'][i]))

Missing Values in event_id are count    0.0
Name: event_id, dtype: float64
Missing Values in device_id are count    453.0
dtype: float64
Missing Values in longitude are count    423.0
dtype: float64
Missing Values in latitude are count    423.0
dtype: float64


## Observations (data.describe()

- Missing values in device_id = 453
- Missing values in longitude = 423
- Missing values in latitude = 423

## Analysis of object data state city

In [19]:
df[['state','city']].nunique()

state     32
city     933
dtype: int64

## Observations
Unique states present - 32<br>
Unique City present - 933

In [20]:
df[['state','city']].isna().sum()

state    377
city       0
dtype: int64

In [70]:
df['state'].value_counts()

Delhi                       751733
Maharashtra                 677105
TamilNadu                   436282
AndhraPradesh               321332
MadhyaPradesh               232690
Rajasthan                   196361
WestBengal                  196161
Karnataka                    99058
UttarPradesh                 94750
Gujarat                      42881
Bihar                        35472
Punjab                       26195
Kerala                       23148
Haryana                      21053
Jharkhand                    18060
Orissa                       15904
Telangana                    12784
Assam                        12021
Chhattisgarh                  9754
Uttaranchal                   7720
JammuandKashmir               5385
Goa                           3250
Nagaland                      2298
Pondicherry                   1741
Mizoram                       1613
Meghalaya                     1515
AndamanandNicobarIslands      1515
HimachalPradesh               1409
Manipur             

Missing values present in state column - 377

In [21]:
#counting the unique values
for columns in df:
  print ("The unique entries in {} is {}". format (columns, df[columns].nunique()))


The unique entries in event_id is 3252950
The unique entries in device_id is 60865
The unique entries in timestamp is 588126
The unique entries in longitude is 60095
The unique entries in latitude is 60177
The unique entries in city is 933
The unique entries in state is 32


In [22]:
#counting the null values
for columns in df:
  print("The null values in {} is {}".format(columns, df[columns].isnull().sum()))


The null values in event_id is 0
The null values in device_id is 453
The null values in timestamp is 0
The null values in longitude is 423
The null values in latitude is 423
The null values in city is 0
The null values in state is 377


# Gender Age data from Mysql Server

In [23]:
# try:
#     mydb = connection.connect(host="cpanel.insaid.co", database = 'Capstone1',user="student", passwd="student",use_pure=True)
#     query = "Select * from gender_age_train;"
#     result_dataFrame = pd.read_sql(query,mydb)
#     mydb.close() #close the connection
# except Exception as e:
#     mydb.close()
#     print(str(e))

In [24]:
# result_dataFrame.to_csv('gender_age.csv')

In [25]:
# df1 = pd.read_csv('gender_age.csv')

# Phone brand device model data from Mysql Server

In [26]:
# try:
#     mydb = connection.connect(host="cpanel.insaid.co", database = 'Capstone1',user="student", passwd="student",use_pure=True)
#     query = "Select * from phone_brand_device_model;"
#     result_dataFrame = pd.read_sql(query,mydb)
#     mydb.close() #close the connection
# except Exception as e:
#     mydb.close()
#     print(str(e))

In [27]:
# result_dataFrame.to_csv('phone_brand_device_model.csv')

In [28]:
# df2 = pd.read_csv('phone_brand_device_model.csv')

In [29]:
# for column in df.columns:
#   print(df[column].unique())
duplicated_list = df.duplicated()

In [30]:
df[duplicated_list]

Empty DataFrame
Columns: [event_id, device_id, timestamp, longitude, latitude, city, state]
Index: []

## Observations
No duplicate entries present in the dataset

# Same analysis for dataset 2 Gender_age.csv

In [58]:
df2 = pd.read_csv('https://raw.githubusercontent.com/ShubhamCJagtap/Capstone_Project_1_EDA/master/gender_age.csv')

In [32]:
df2.head()

Unnamed: 0            device_id gender  age   group
0           0 -8076087639492063270      M   35  M32-38
1           1 -2897161552818060146      M   35  M32-38
2           2 -8260683887967679142      M   35  M32-38
3           3 -4938849341048082022      M   30  M29-31
4           4   245133531816851882      M   30  M29-31

In [33]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 74645 entries, 0 to 74644
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Unnamed: 0  74645 non-null  int64 
 1   device_id   74645 non-null  int64 
 2   gender      74645 non-null  object
 3   age         74645 non-null  int64 
 4   group       74645 non-null  object
dtypes: int64(3), object(2)
memory usage: 2.8+ MB


# Observations

Total Entries - 74646<br>
No Null values detected<br>
More analysis required for group column

In [34]:
df2.describe()

Unnamed: 0     device_id           age
count  74645.000000  7.464500e+04  74645.000000
mean   37322.000000 -7.491354e+14     31.410342
std    21548.299759  5.327150e+18      9.868735
min        0.000000 -9.223067e+18      1.000000
25%    18661.000000 -4.617367e+18     25.000000
50%    37322.000000 -1.841362e+16     29.000000
75%    55983.000000  4.636656e+18     36.000000
max    74644.000000  9.222849e+18     96.000000

## Observation gender_age
age minimum - 1 , maximum - 96 and mean = 31.41 , median = 29<br>
Mean and median don't have high difference <br>
Assumption check - age is normally distributed ??

In [35]:
df2['gender'].value_counts()

M    47904
F    26741
Name: gender, dtype: int64

## Observation

male entries are 47904<br>
female entries are 26741<br>

In [36]:
df2.nunique()

Unnamed: 0    74645
device_id     74645
gender            2
age              85
group            12
dtype: int64

In [37]:
df2['group'].unique()

array(['M32-38', 'M29-31', 'F24-26', 'F33-42', 'F27-28', 'M39+', 'M23-26',
       'M27-28', 'M22-', 'F43+', 'F23-', 'F29-32'], dtype=object)

In [38]:
print("Male above age 39 are {}".format(df2[(df2['age']>39)&(df2['gender']=='M')]['age'].sum()))

print("Female above age 43 are {}".format(df2[(df2['age']>43)&(df2['gender']=='F')]['age'].sum()))


Male above age 39 are 372416
Female above age 43 are 199865


In [39]:
print("Male below age 22 are {}".format(df2[(df2['age']<22)&(df2['gender']=='M')]['age'].sum()))
print("Female below age 23 are {}".format(df2[(df2['age']<23)&(df2['gender']=='F')]['age'].sum()))

Male below age 22 are 108781
Female below age 23 are 80571


In [40]:
print("Male below age 10 are {}".format(df2[(df2['age']<10)&(df2['gender']=='M')]['age'].sum()))
print("Female below age 10 are {}".format(df2[(df2['age']<10)&(df2['gender']=='F')]['age'].sum()))

Male below age 10 are 25
Female below age 10 are 0


In [41]:
print("Male below age 5 are {}".format(df2[(df2['age']<5)&(df2['gender']=='M')]['age'].sum()))
print("Female below age 5 are {}".format(df2[(df2['age']<5)&(df2['gender']=='F')]['age'].sum()))

Male below age 5 are 1
Female below age 5 are 0


In [42]:
df2.sort_values('age',ascending=True)[['age','group']].head(25)

age group
10385    1  M22-
60833    6  M22-
60832    6  M22-
60779    6  M22-
60831    6  M22-
57703   10  M22-
33595   10  M22-
15107   10  M22-
72936   10  F23-
72935   10  F23-
53296   11  F23-
53282   11  F23-
53994   11  F23-
52335   11  F23-
54609   12  M22-
53603   12  M22-
60527   12  M22-
61136   12  F23-
53642   12  F23-
59192   12  F23-
58421   12  M22-
61398   12  M22-
53136   12  F23-
55195   12  M22-
11510   12  M22-

In [43]:
df2.sort_values('age',ascending=False)[['age','group']].head(25)

age group
5958    96  F43+
60872   94  F43+
3704    90  M39+
32290   89  F43+
2971    89  F43+
32291   89  F43+
3068    88  F43+
3560    88  F43+
19602   88  M39+
3065    88  F43+
3067    88  M39+
2972    87  M39+
49530   87  M39+
2973    87  M39+
43127   86  M39+
43128   86  M39+
60342   86  M39+
60349   86  F43+
27797   85  M39+
60348   85  M39+
37118   85  M39+
3573    84  M39+
2799    84  F43+
9150    83  M39+
9151    83  M39+

## identified Question
What should be the age range to be considered in EDA? 

# Unique Values
Gender - 2 (M,F)<br>
age - 85<br>
group - 12<br>

#groups are present as per gender

Male age group starts from M22- till M39+<br>
Female age group starts from F23- till F43+<br>

In [44]:
duplicated_list_df2 = df2.drop('Unnamed: 0',axis=1).duplicated()

In [45]:
df2.drop('Unnamed: 0',axis=1)[duplicated_list_df2]

Empty DataFrame
Columns: [device_id, gender, age, group]
Index: []

#Analysis of Phone_brand_dataset

In [46]:
df3 = pd.read_csv('https://raw.githubusercontent.com/ShubhamCJagtap/Capstone_Project_1_EDA/master/phone_brand_device_model.csv')

In [47]:
df3.head(10)

Unnamed: 0            device_id phone_brand        device_model
0           0  1877775838486905855        vivo                 Y13
1           1 -3766087376657242966          小米                V183
2           2 -6238937574958215831        OPPO                 R7s
3           3  8973197758510677470          三星               A368t
4           4 -2015528097870762664          小米             红米Note2
5           5 -8896952452425303597          小米             红米Note3
6           6 -2634171391322875113          小米                S5.1
7           7 -2526386806972133373          三星  Galaxy Grand Prime
8           8  1427934937997164515          小米               MI 2S
9           9  2866484947180243791          三星       Galaxy Note 2

In [48]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 87726 entries, 0 to 87725
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Unnamed: 0    87726 non-null  int64 
 1   device_id     87726 non-null  int64 
 2   phone_brand   87726 non-null  object
 3   device_model  87726 non-null  object
dtypes: int64(2), object(2)
memory usage: 2.7+ MB


## Observation
Total entries - 87726<br>
No missing values

In [49]:
df3.nunique()

Unnamed: 0      87726
device_id       87726
phone_brand       116
device_model     1467
dtype: int64

In [50]:
df3['phone_brand'].unique()

array(['vivo', '小米', 'OPPO', '三星', '酷派', '联想 ', '华为', '奇酷', '魅族', '斐讯',
       '中国移动', 'HTC', '天语', '至尊宝', 'LG', '欧博信', '优米', 'ZUK', '努比亚', '惠普',
       '尼比鲁', '美图', '乡米', '摩托罗拉', '梦米', '锤子', '富可视', '乐视', '海信', '百立丰',
       '一加', '语信', '海尔', '酷比', '纽曼', '波导', '朵唯', '聆韵', 'TCL', '酷珀', '爱派尔',
       'LOGO', '青葱', '果米', '华硕', '昂达', '艾优尼', '康佳', '优购', '邦华', '赛博宇华',
       '黑米', 'Lovme', '先锋', 'E派', '神舟', '诺基亚', '普耐尔', '糖葫芦', '亿通', '欧新',
       '米奇', '酷比魔方', '蓝魔', '小杨树', '贝尔丰', '糯米', '米歌', 'E人E本', '西米', '大Q',
       '台电', '飞利浦', '唯米', '大显', '长虹', '维图', '青橙', '本为', '虾米', '夏新', '帷幄',
       '百加', 'SUGAR', '欧奇', '世纪星', '智镁', '欧比', '基伍', '飞秒', '德赛', '易派',
       '谷歌', '金星数码', '广信', '诺亚信', 'MIL', '白米', '大可乐', '宝捷讯', '优语', '首云',
       '瑞米', '瑞高', '沃普丰', '摩乐', '鲜米', '凯利通', '唯比', '欧沃', '丰米', '恒宇丰',
       '奥克斯', '西门子', '欧乐迪', 'PPTV'], dtype=object)

# Observation
phone_brand have many chinese brand names and company provided equivalent english names are limited.<br>


Brand Name	Brand English Mapping<br><br>
'华为'	'Huawei'<br>
'小米'	'Xiaomi'<br>
'三星'	'Samsung'<br>
'vivo'	'vivo'<br>
'OPPO'	'OPPO'<br>
'魅族'	'Meizu'<br>
'酷派'	'Coolpad'<br>
'乐视'	'LeEco'<br>
'联想 '	'Lenovo'<br>
'HTC'	'HTC'<br><br>

We shall check these brand names for Delhi only


In [51]:
for i in df3['device_model'].unique():
  print(i,end='   ')

Y13   V183   R7s   A368t   红米Note2   红米Note3   S5.1   Galaxy Grand Prime   MI 2S   Galaxy Note 2   R1C   大神F1Plus   note顶配版   Galaxy Mega 2   星星1号   Galaxy S4   MI 3   Find 7   C8816   红米note   X6 L   Galaxy Win   青春版   MX3   Y35   G520-5000   荣耀4A   魅蓝Note 2   Y560-CL00   荣耀7i   3000   荣耀畅玩4C   C730Lw   红米2A   A1   One   Galaxy J7   R3   MX2   荣耀畅玩5   荣耀7   MI 4   One M8   红米1S   麦芒4   Galaxy S6 Edge   A3800d   P780   Ascend G7   荣耀6   Xperia SL   Y17T   荣耀畅玩4X   Ascend P7   Galaxy Grand 2   荣耀3X畅玩版   荣耀X2   Ascend P8   R2017   魅蓝NOTE   Mate 2   M8   T87+   X710L   A890e   R7   X5M   MX5   Galaxy Mega 5.8   L81   Xperia Z1   荣耀6 Plus   X5Max+   Galaxy J5   Galaxy Note 4   A31   GN706L   G3   P8 Lite   Sensation XE with Beats Audio Z715e   A880   荣耀+   X5L   8720L   IVO 6655   R7 Plus   荣耀畅玩4   MX4   Y635-TL00   Galaxy S3   X3L   3   R807   黄金斗士A8   One X   小米note   R6007   Galaxy S2 Plus   R1   荣耀3C   小米4C   R823T   Mate 7   MI 2   F2   Galaxy S5   Xperia TX   Galaxy Trend DUOS 2   Z1

## Observation
In device model, few names are in chinese. <br>
No equivalent english names provided

In [52]:
duplicate_in_df3 = df3.drop('Unnamed: 0',axis=1).duplicated()

In [53]:
df3.drop('Unnamed: 0',axis=1)[duplicate_in_df3]

Empty DataFrame
Columns: [device_id, phone_brand, device_model]
Index: []

No Duplicate values present

In [54]:
# id = []
# miss = []
# for i in df2['device_id'].unique():
#   if i in df['device_id'].unique():
#     id.append(i)
#   else:
#     miss.append(i)

KeyboardInterrupt: ignored

In [55]:
# df2.drop('Unnamed: 0',axis=1,inplace=True)

In [59]:
# merge all tables by performing outer join between events, gender and phone brand
merged_1 = df.merge(df2.drop('Unnamed: 0',axis=1), how = "left", on = "device_id")
# .merge(phonebrand_device_model_df,how = "outer", on = "device_id",suffixes=('_left', '_right'))


In [60]:
merged_1.head()

event_id     device_id            timestamp  longitude   latitude  \
0   2765368  2.973348e+18  2016-05-07 22:52:05  77.225676  28.730140   
1   2955066  4.734221e+18  2016-05-01 20:44:16  88.388361  22.660325   
2    605968 -3.264500e+18  2016-05-02 14:23:04  77.256809  28.757906   
3    448114  5.731369e+18  2016-05-03 13:21:16  80.343613  13.153332   
4    665740  3.388880e+17  2016-05-06 03:51:05  85.997745  23.842609   

       city       state gender   age   group  
0     Delhi       Delhi      M  35.0  M32-38  
1  Calcutta  WestBengal    NaN   NaN     NaN  
2     Delhi       Delhi    NaN   NaN     NaN  
3   Chennai   TamilNadu    NaN   NaN     NaN  
4    Bokaro   Jharkhand    NaN   NaN     NaN

In [62]:
merged_1.shape

(3252950, 10)

In [63]:
merged_1.isna().sum()

event_id           0
device_id        453
timestamp          0
longitude        423
latitude         423
city               0
state            377
gender       3235968
age          3235968
group        3235968
dtype: int64

In [64]:
df.nunique()

event_id     3252950
device_id      60865
timestamp     588126
longitude      60095
latitude       60177
city             933
state             32
dtype: int64

In [65]:
df2.nunique()

Unnamed: 0    74645
device_id     74645
gender            2
age              85
group            12
dtype: int64

In [66]:
merged_1.nunique()

event_id     3252950
device_id      60865
timestamp     588126
longitude      60095
latitude       60177
city             933
state             32
gender             2
age               51
group             12
dtype: int64

In [67]:
merged_2 = merged_1.merge(df3.drop('Unnamed: 0',axis=1), how = "left", on = "device_id")

In [68]:
merged_2.nunique()

event_id        3252950
device_id         60865
timestamp        588126
longitude         60095
latitude          60177
city                933
state                32
gender                2
age                  51
group                12
phone_brand          26
device_model        191
dtype: int64

In [69]:
merged_2.isna().sum()

event_id              0
device_id           453
timestamp             0
longitude           423
latitude            423
city                  0
state               377
gender          3235968
age             3235968
group           3235968
phone_brand     3235968
device_model    3235968
dtype: int64

In [71]:
merged_df_df2 = df.merge(df2.drop('Unnamed: 0',axis=1), how = "inner", on = "device_id")

In [72]:
merged_df_df2.nunique()

event_id     16982
device_id      406
timestamp    16553
longitude      406
latitude       406
city           197
state           26
gender           2
age             51
group           12
dtype: int64

In [73]:
merged_df_df3 = df.merge(df3.drop('Unnamed: 0',axis=1), how = "inner", on = "device_id")

In [74]:
df[df['state']=='Delhi'].nunique()

event_id     751733
device_id      4909
timestamp    412084
longitude      4795
latitude       4789
city             10
state             1
dtype: int64

In [75]:
merged_df_df3.nunique()

event_id        16982
device_id         406
timestamp       16553
longitude         406
latitude          406
city              197
state              26
phone_brand        26
device_model      191
dtype: int64

In [76]:
merged_df2_df3 = df2.merge(df3.drop('Unnamed: 0',axis=1), how = "inner", on = "device_id")

In [77]:
merged_df2_df3.nunique()

Unnamed: 0      74617
device_id       74617
gender              2
age                85
group              12
phone_brand       111
device_model     1427
dtype: int64

In [79]:
all = merged_df2_df3.merge(df, how = "inner", on = "device_id")

In [80]:
all.nunique()

Unnamed: 0        406
device_id         406
gender              2
age                51
group              12
phone_brand        26
device_model      191
event_id        16982
timestamp       16553
longitude         406
latitude          406
city              197
state              26
dtype: int64

In [83]:
all[all['state']=='Delhi'].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2328 entries, 1059 to 16813
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Unnamed: 0    2328 non-null   int64  
 1   device_id     2328 non-null   int64  
 2   gender        2328 non-null   object 
 3   age           2328 non-null   int64  
 4   group         2328 non-null   object 
 5   phone_brand   2328 non-null   object 
 6   device_model  2328 non-null   object 
 7   event_id      2328 non-null   int64  
 8   timestamp     2328 non-null   object 
 9   longitude     2328 non-null   float64
 10  latitude      2328 non-null   float64
 11  city          2328 non-null   object 
 12  state         2328 non-null   object 
dtypes: float64(2), int64(4), object(7)
memory usage: 254.6+ KB
